### DISCUSSION TITLE
Esp 32 midi controller

### DISCUSSION THREAD

**Discussion by KingMichaelCoy** [tttapa/Control-Surface#793] [General]
What is the best was to make something like

  [](https://learn.adafruit.com/web-ble-midi/ble-midi-controller)

But for a esp32 or Adafruit huzzah 32?

**Comment by josehdx** [comment]
What do you want to build? Your question is not clear to me. A Bluetooth midi device? An usb-midi ?

**Reply by KingMichaelCoy** [reply]
A Bluetooth midi controller with knobs potentiometers and push button or arcade buttons to play notes with in a daw


[](https://learn.adafruit.com/web-ble-midi/ble-midi-controller)

**Reply by josehdx** [reply]
You need a esp 32. The library only support Bluetooth with esp32 boards

**Reply by KingMichaelCoy** [reply]
[]() of course I have one

**Reply by KingMichaelCoy** [reply]
I wanted to know how to build a code from one or modify the one from the link on my original post

**Reply by josehdx** [reply]
> 
Fro arcade buttons, you can use "MIDIAddress" and assign te pin number from your board you want to user, for potentiometers there is one class as well, but i have not used it so far.

This is the code i am using for buttons:

`#include <Control_Surface.h>
BluetoothMIDI_Interface midi;


CCButton button[] {
  {0, {MIDI_CC::General_Purpose_Controller_1, CHANNEL_1} },
  {35, {MIDI_CC::General_Purpose_Controller_1, CHANNEL_2} },
};

void setup() {
  Control_Surface.begin(); // Initialize Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface
}`


Where "BluetoothMIDI_Interface midi;" sets my ESP32 as BLE MIDI and the 0 and 35 on " {0, {MIDI_CC::General_Purpose_Controller_1, CHANNEL_1} }, {35, {MIDI_CC::General_Purpose_Controller_1, CHANNEL_2} }," are the pin numbers on my BLE. You need to know where you want to solder such buttons and then assign it on the code

**Comment by KingMichaelCoy** [comment]
Ok, I have the code working with my esp32, well,, its actually a Adafruit feather huzzah 32 which is a esp32 but I also have a esp32 that is a Microcenter brand/version. Additionally I have a Adafruit qtpy esp32 c3 which does not have native usb but all of these microcontrollers have the esp32 so I would use these for bluetooth midi with this sketch. However, in getting the code to be uploaded successfully and my device connected to my daw, the pins I assigned were 34 and 12. I am not getting any midi signal that is triggering the corresponding sounds on my daw. Am I doing something incorrectly?

**Reply by josehdx** [reply]
So the BLE Connection is ok? Do you have a Windows pc?  I recommend you to try first this android app (https://play.google.com/store/apps/details?id=dcm.pianomidibleusb) dont pair your board via android system. This app will be able to pair and monitor your ble. Try it and check if you have midi messages going out of your board

**Reply by KingMichaelCoy** [reply]
I have Mac and a Linux pi4

**Reply by josehdx** [reply]
> I have Mac and a Linux pi4

Ir might be something with your daw. Try to use any MIDI Monitoring tool. If you are setting the pins correctly then it should work i guess

**Reply by KingMichaelCoy** [reply]
I'm going to try in Zenbeats for iPhone but I am using the 12 and 34 pins in the feather huzzah 32

**Reply by KingMichaelCoy** [reply]
Is this code for any type of esp32? Like Adafruit, c3 s2 etc?

**Reply by josehdx** [reply]
Yes. It should work with any esp32

**Reply by KingMichaelCoy** [reply]
So if my code is uploaded and pins connected then my arcade button should trigger?

**Reply by KingMichaelCoy** [reply]
I can add pictures.  Shortly on my first rest stop.

**Reply by KingMichaelCoy** [reply]
https://user-images.githubusercontent.com/102106403/188289016-82bc3c80-6858-4089-9a9e-66ed87d16e76.MOV

**Reply by KingMichaelCoy** [reply]
This is what I built with the nrf52 and I know how to hook my arcade buttons with soildered jst connectors. I'm just trying to get this working with the Adafruit feather huzzah 32 but I feel like I will get it. Also adding a apds 9960 to a qtpy esp32 c3 to have a wireless gesture modulated midi controller

**Reply by KingMichaelCoy** [reply]
So I got a midi monitor for my iPhone.  Yes it sends signal through channel one. But no sound triggering

**Reply by josehdx** [reply]
ok now i got it, you are sending the wrong MIDI CC message in that case. but I can not really help further because i dont really send MIDI notes but other kind of data like the one in my code, you need to check the documentation to see how to send notes. sorry for not helping further

**Reply by KingMichaelCoy** [reply]
Pin 22 for channel 1and 1 for channel 2

It shows signal sending but does not write or trigger the data

**Reply by KingMichaelCoy** [reply]
Thank you for helping me get to this point. I'll look at the documentation. Is there a way to show support for your help? What exactlyI'm the documentation would I look at

**Reply by josehdx** [reply]
Probably the answer of the developer on this link will help:

https://github.com/tttapa/Control-Surface/discussions/527

**Reply by KingMichaelCoy** [reply]
I lost the initial code you sent can you direct me to the initial thread I got the midi notes working but not programmed? It was a code you sent the only one in the thread. However I can't find it in my GitHub mobile thread. 

Maybe combine the two codes


Is there a way to schedule a tutoring with you and actually know how to write basic. I understand why it work but I want to be like macguyver and do something with magnesium and a iron and trinket?

Going to the link. Thank you

**Comment by josehdx** [comment]
yeah, because probably you are sending the worng Data, MIDI can be many thins, notes, pitch shiftint etc. Or maybe there is a setting on the app that is not writing the Data you are sending, but if the BLE connection and the MIDI sending is working, is more like a setup thing. Double check how to send MIDI notes on the documentation, i can not really do more. Sorry

**Comment by Veroledez** [comment]
Hola, para enviar notas midi necesitas usar lo siguiente


NoteButton button {
22,    // aqui el pin que vas a usar
{MIDI_Notes::C(4), CHANNEL_1}, // nota C4 en el canal midi 1
127,   //  la velocidad de la nota, ya que es fija, no varia independientemente de la fuerza con que presiones el boton, para velocidad variable hay que usar un fsr o dos botones, y hay que crear una clase que lo soporte, la que menciono aqui es solo velocidad fija
};


Si tienen algun problema, publica de forma enumerada tus preguntas y postea el codigo que tienes para ayudarte

**Comment by KingMichaelCoy** [comment]
I got it working but only button 22. I want my other five buttons working two knobs and a joystick on a 1602a
